src/main/scala/uk/gov/hmrc/ct/version/calculations/ReturnVersionsCalculator.scala

```diff
@@ -35,62 +35,61 @@ object ReturnVersionsCalculator extends ReturnVersionsCalculator
 
 trait ReturnVersionsCalculator {
 
-  def doCalculation[A <: BoxRetriever](boxRetriever: A): Set[Return] = {
-    boxRetriever match {
+  def doCalculation[A <: BoxRetriever](filingAttributesBoxValueRetriever: FilingAttributesBoxValueRetriever,
+                                       boxRetriever: Option[A]): Set[Return] = {
+    (filingAttributesBoxValueRetriever, boxRetriever) match {
 
-      case br: V3CT600EBoxRetriever with AccountsBoxRetriever with FilingAttributesBoxValueRetriever =>
+      case (baseBoxRetriever, Some(br: V3CT600EBoxRetriever with AccountsBoxRetriever)) =>
         calculateReturnVersions(poaStartDate = br.ac3().value,
                                 apStartDate = Some(br.e3().value),
                                 apEndDate = Some(br.e4().value),
-                                coHoFiling = br.companiesHouseFiling(),
-                                hmrcFiling = br.hmrcFiling(),
-                                microEntityFiling = br.microEntityFiling(),
-                                statutoryAccountsFiling = br.statutoryAccountsFiling(),
-                                abridgedFiling = br.abridgedFiling(),
-                                abbreviatedAccountsFiling = br.abbreviatedAccountsFiling(),
-                                companyType = br.companyType(),
+                                coHoFiling = baseBoxRetriever.companiesHouseFiling(),
```

Comment:
The filing boxes look repeated here, and then we're just calculating the two charity boxes, so could we use the functional style like the code above in two functions for `charityAllExempt` and `charityNoIncome` values and then remove all the case statements? That would be pretty awesome.

Reply:
I'll take a look later